Crash on uplod failure

**Summary:** 
I can consistently reproduce a crash on current master by making the upload fail intentionally. You can disconnect yourself during the upload to reproduce, upload will fail as expected and app crashes (this one is unexpected). Here are whole logs I could collect on verbose mode.

Tested on API 26 emulator with current master.
```
   <?xml version="1.0" encoding="UTF-8"?><api batchcomplete=""><query><normalized><n from="t.jpg" to="T.jpg"/></normalized><pages><page _idx="-1" ns="0" title="T.jpg" missing=""/></pages></query></api>
2019-03-21 12:48:38.432 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< START OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:38.432 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Request line:
     GET /w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml HTTP/1.1
2019-03-21 12:48:38.432 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:38.433 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:38.433 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:38.433 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Host: commons.wikimedia.beta.wmflabs.org
2019-03-21 12:48:38.433 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Connection: Keep-Alive
2019-03-21 12:48:38.433 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> User-Agent: Commons/2.10.1-debug-master~8cd87ad14 (https://mediawiki.org/wiki/Apps/Commons) Android/8.0.0
2019-03-21 12:48:38.434 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie: WMF-Last-Access=21-Mar-2019; WMF-Last-Access-Global=21-Mar-2019; GeoIP=TR:06:Ankara:39.91:32.84:v4; commonswikiSession=ed772s8lsha0amt4ldo4crd2umevgnbv; commonswikiUserID=1694; commonswikiUserName=Neslihan+Test; centralauth_User=Neslihan+Test; centralauth_Token=3e9640fadabac6a356f65a62f989a103; centralauth_Session=4e49ed09ffff1e7626f066baf9cacbf6; loginnotify_prevlogins=2019-1fok289-sszbcgr3m09htbifi12fmr8hpow63eu
2019-03-21 12:48:38.434 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie2: $Version=1
2019-03-21 12:48:38.434 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:38.434 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Protocol version:
     HTTP/1.1
2019-03-21 12:48:38.435 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< END OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:38.504 20467-20467/fr.free.nrw.commons.beta D/NearbyNotificationCardView: Update nearby card notification content
2019-03-21 12:48:38.529 20467-20472/fr.free.nrw.commons.beta I/zygote: Do partial code cache collection, code=992KB, data=579KB
2019-03-21 12:48:38.529 20467-20472/fr.free.nrw.commons.beta I/zygote: After code cache collection, code=992KB, data=579KB
2019-03-21 12:48:38.529 20467-20472/fr.free.nrw.commons.beta I/zygote: Increasing code cache capacity to 3MB
2019-03-21 12:48:38.662 20467-20601/fr.free.nrw.commons.beta D/CustomApiResult: API response for method https://commons.wikimedia.beta.wmflabs.org/w/api.php is
     <?xml version="1.0" encoding="UTF-8"?><api batchcomplete=""><query><normalized><n from="t.jpg" to="T.jpg"/></normalized><pages><page _idx="-1" ns="0" title="T.jpg" missing=""/></pages></query></api>
2019-03-21 12:48:38.684 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< START OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:38.685 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Request line:
     GET /w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml HTTP/1.1
2019-03-21 12:48:38.685 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:38.685 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:38.685 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:38.685 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Host: commons.wikimedia.beta.wmflabs.org
2019-03-21 12:48:38.686 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Connection: Keep-Alive
2019-03-21 12:48:38.686 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> User-Agent: Commons/2.10.1-debug-master~8cd87ad14 (https://mediawiki.org/wiki/Apps/Commons) Android/8.0.0
2019-03-21 12:48:38.686 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie: WMF-Last-Access=21-Mar-2019; WMF-Last-Access-Global=21-Mar-2019; GeoIP=TR:06:Ankara:39.91:32.84:v4; commonswikiSession=ed772s8lsha0amt4ldo4crd2umevgnbv; commonswikiUserID=1694; commonswikiUserName=Neslihan+Test; centralauth_User=Neslihan+Test; centralauth_Token=3e9640fadabac6a356f65a62f989a103; centralauth_Session=4e49ed09ffff1e7626f066baf9cacbf6; loginnotify_prevlogins=2019-1fok289-sszbcgr3m09htbifi12fmr8hpow63eu
2019-03-21 12:48:38.686 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie2: $Version=1
2019-03-21 12:48:38.686 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:38.687 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Protocol version:
     HTTP/1.1
2019-03-21 12:48:38.687 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< END OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
    
    --------- beginning of system
2019-03-21 12:48:38.788 1680-1680/system_process W/WindowManager: removeWindowToken: Attempted to remove non-existing token: android.os.Binder@7efa9c1
2019-03-21 12:48:38.920 20467-20601/fr.free.nrw.commons.beta D/CustomApiResult: API response for method https://commons.wikimedia.beta.wmflabs.org/w/api.php is
     <?xml version="1.0" encoding="UTF-8"?><api batchcomplete=""><query><normalized><n from="t.jpg" to="T.jpg"/></normalized><pages><page _idx="-1" ns="0" title="T.jpg" missing=""/></pages></query></api>
2019-03-21 12:48:38.942 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< START OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:38.942 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Request line:
     GET /w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml HTTP/1.1
2019-03-21 12:48:38.942 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:38.942 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:38.943 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:38.943 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Host: commons.wikimedia.beta.wmflabs.org
2019-03-21 12:48:38.943 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Connection: Keep-Alive
2019-03-21 12:48:38.943 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> User-Agent: Commons/2.10.1-debug-master~8cd87ad14 (https://mediawiki.org/wiki/Apps/Commons) Android/8.0.0
2019-03-21 12:48:38.944 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie: WMF-Last-Access=21-Mar-2019; WMF-Last-Access-Global=21-Mar-2019; GeoIP=TR:06:Ankara:39.91:32.84:v4; commonswikiSession=ed772s8lsha0amt4ldo4crd2umevgnbv; commonswikiUserID=1694; commonswikiUserName=Neslihan+Test; centralauth_User=Neslihan+Test; centralauth_Token=3e9640fadabac6a356f65a62f989a103; centralauth_Session=4e49ed09ffff1e7626f066baf9cacbf6; loginnotify_prevlogins=2019-1fok289-sszbcgr3m09htbifi12fmr8hpow63eu
2019-03-21 12:48:38.944 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie2: $Version=1
2019-03-21 12:48:38.944 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:38.944 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Protocol version:
     HTTP/1.1
2019-03-21 12:48:38.944 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< END OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:38.995 1435-5720/? W/audio_hw_generic: Not supplying enough data to HAL, expected position 24757934 , only wrote 24757920
2019-03-21 12:48:39.146 20467-20467/fr.free.nrw.commons.beta I/Choreographer: Skipped 31 frames!  The application may be doing too much work on its main thread.
2019-03-21 12:48:39.231 1786-1786/com.android.systemui D/StatusBar: disable<e i a s b h r c s q >
2019-03-21 12:48:39.311 1786-2049/com.android.systemui D/EGL_emulation: eglMakeCurrent: 0x8d51b820: ver 2 0 (tinfo 0x9ce31e40)
2019-03-21 12:48:39.314 1441-1441/? D/gralloc_ranchu: gralloc_alloc: Creating ashmem region of size 14749696
2019-03-21 12:48:39.345 1786-2049/com.android.systemui D/EGL_emulation: eglMakeCurrent: 0x8d51b820: ver 2 0 (tinfo 0x9ce31e40)
2019-03-21 12:48:39.346 20467-20601/fr.free.nrw.commons.beta D/CustomApiResult: API response for method https://commons.wikimedia.beta.wmflabs.org/w/api.php is
     <?xml version="1.0" encoding="UTF-8"?><api batchcomplete=""><query><normalized><n from="t.jpg" to="T.jpg"/></normalized><pages><page _idx="-1" ns="0" title="T.jpg" missing=""/></pages></query></api>
2019-03-21 12:48:39.347 1786-1786/com.android.systemui D/StatusBar: disable<e i a s b h r c s q >
2019-03-21 12:48:39.401 1441-1441/? D/gralloc_ranchu: gralloc_alloc: Creating ashmem region of size 14749696
2019-03-21 12:48:39.468 1786-1786/com.android.systemui I/zygote: Deoptimizing void com.android.systemui.statusbar.stack.ViewState.startYTranslationAnimation(android.view.View, com.android.systemui.statusbar.stack.AnimationProperties) due to JIT inline cache
2019-03-21 12:48:39.567 1441-1441/? D/gralloc_ranchu: gralloc_alloc: Creating ashmem region of size 14749696
2019-03-21 12:48:40.121 1435-1570/? W/audio_hw_generic: Not supplying enough data to HAL, expected position 24865716 , only wrote 24811920
2019-03-21 12:48:40.343 20467-20498/fr.free.nrw.commons.beta D/UploadService$NotificationUpdateProgressListener: Uploaded 1127634 of 7492596
2019-03-21 12:48:40.398 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< START OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:40.398 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: Request line:
     GET /w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml HTTP/1.1
2019-03-21 12:48:40.398 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:40.399 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:40.399 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:40.400 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Host: commons.wikimedia.beta.wmflabs.org
2019-03-21 12:48:40.400 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Connection: Keep-Alive
2019-03-21 12:48:40.400 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> User-Agent: Commons/2.10.1-debug-master~8cd87ad14 (https://mediawiki.org/wiki/Apps/Commons) Android/8.0.0
2019-03-21 12:48:40.401 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie: WMF-Last-Access=21-Mar-2019; WMF-Last-Access-Global=21-Mar-2019; GeoIP=TR:06:Ankara:39.91:32.84:v4; commonswikiSession=ed772s8lsha0amt4ldo4crd2umevgnbv; commonswikiUserID=1694; commonswikiUserName=Neslihan+Test; centralauth_User=Neslihan+Test; centralauth_Token=3e9640fadabac6a356f65a62f989a103; centralauth_Session=4e49ed09ffff1e7626f066baf9cacbf6; loginnotify_prevlogins=2019-1fok289-sszbcgr3m09htbifi12fmr8hpow63eu
2019-03-21 12:48:40.401 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie2: $Version=1
2019-03-21 12:48:40.404 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:40.405 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: Protocol version:
     HTTP/1.1
2019-03-21 12:48:40.405 20467-20600/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< END OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:40.635 20467-20600/fr.free.nrw.commons.beta D/CustomApiResult: API response for method https://commons.wikimedia.beta.wmflabs.org/w/api.php is
     <?xml version="1.0" encoding="UTF-8"?><api batchcomplete=""><query><normalized><n from="t.jpg" to="T.jpg"/></normalized><pages><page _idx="-1" ns="0" title="T.jpg" missing=""/></pages></query></api>
2019-03-21 12:48:40.658 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< START OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:40.675 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Request line:
     GET /w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml HTTP/1.1
2019-03-21 12:48:40.675 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:40.675 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:40.676 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:40.676 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Host: commons.wikimedia.beta.wmflabs.org
2019-03-21 12:48:40.676 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Connection: Keep-Alive
2019-03-21 12:48:40.677 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> User-Agent: Commons/2.10.1-debug-master~8cd87ad14 (https://mediawiki.org/wiki/Apps/Commons) Android/8.0.0
2019-03-21 12:48:40.677 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie: WMF-Last-Access=21-Mar-2019; WMF-Last-Access-Global=21-Mar-2019; GeoIP=TR:06:Ankara:39.91:32.84:v4; commonswikiSession=ed772s8lsha0amt4ldo4crd2umevgnbv; commonswikiUserID=1694; commonswikiUserName=Neslihan+Test; centralauth_User=Neslihan+Test; centralauth_Token=3e9640fadabac6a356f65a62f989a103; centralauth_Session=4e49ed09ffff1e7626f066baf9cacbf6; loginnotify_prevlogins=2019-1fok289-sszbcgr3m09htbifi12fmr8hpow63eu
2019-03-21 12:48:40.677 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie2: $Version=1
2019-03-21 12:48:40.677 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:40.678 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Protocol version:
     HTTP/1.1
2019-03-21 12:48:40.678 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< END OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:40.934 20467-20601/fr.free.nrw.commons.beta D/CustomApiResult: API response for method https://commons.wikimedia.beta.wmflabs.org/w/api.php is
     <?xml version="1.0" encoding="UTF-8"?><api batchcomplete=""><query><normalized><n from="t.jpg" to="T.jpg"/></normalized><pages><page _idx="-1" ns="0" title="T.jpg" missing=""/></pages></query></api>
2019-03-21 12:48:40.944 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< START OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:40.944 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Request line:
     GET /w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml HTTP/1.1
2019-03-21 12:48:40.945 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:40.945 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< REQUEST PARAMS >>>>>>>>>>>>
2019-03-21 12:48:40.945 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:40.945 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Host: commons.wikimedia.beta.wmflabs.org
2019-03-21 12:48:40.945 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Connection: Keep-Alive
2019-03-21 12:48:40.946 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> User-Agent: Commons/2.10.1-debug-master~8cd87ad14 (https://mediawiki.org/wiki/Apps/Commons) Android/8.0.0
2019-03-21 12:48:40.946 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie: WMF-Last-Access=21-Mar-2019; WMF-Last-Access-Global=21-Mar-2019; GeoIP=TR:06:Ankara:39.91:32.84:v4; commonswikiSession=ed772s8lsha0amt4ldo4crd2umevgnbv; commonswikiUserID=1694; commonswikiUserName=Neslihan+Test; centralauth_User=Neslihan+Test; centralauth_Token=3e9640fadabac6a356f65a62f989a103; centralauth_Session=4e49ed09ffff1e7626f066baf9cacbf6; loginnotify_prevlogins=2019-1fok289-sszbcgr3m09htbifi12fmr8hpow63eu
2019-03-21 12:48:40.946 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Header >> Cookie2: $Version=1
2019-03-21 12:48:40.946 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< HEADERS >>>>>>>>>>>>
2019-03-21 12:48:40.947 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: Protocol version:
     HTTP/1.1
2019-03-21 12:48:40.947 20467-20601/fr.free.nrw.commons.beta V/NetworkInterceptors: <<<<<<<<<<<<<< END OF REQUEST LOGGING [/w/api.php?prop=imageinfo&format=xml&action=query&iiprop=url&titles=t.jpg&iiurlwidth=640&format=xml] >>>>>>>>>>>>
2019-03-21 12:48:41.168 20467-20601/fr.free.nrw.commons.beta D/CustomApiResult: API response for method https://commons.wikimedia.beta.wmflabs.org/w/api.php is
     <?xml version="1.0" encoding="UTF-8"?><api batchcomplete=""><query><normalized><n from="t.jpg" to="T.jpg"/></normalized><pages><page _idx="-1" ns="0" title="T.jpg" missing=""/></pages></query></api>
2019-03-21 12:48:41.522 1680-1691/system_process D/WifiService: setWifiEnabled: false pid=1786, uid=10028, package=com.android.systemui
2019-03-21 12:48:41.530 1680-1756/system_process D/WifiStateMachine: WifiStateMachine: Leaving Connected state
2019-03-21 12:48:41.531 1680-1757/system_process D/DhcpClient: doQuit
2019-03-21 12:48:41.533 1680-20543/system_process D/DhcpClient: onQuitting
2019-03-21 12:48:41.535 1539-1660/? D/CommandListener: Setting iface cfg
2019-03-21 12:48:41.538 1435-5720/? W/audio_hw_generic: Not supplying enough data to HAL, expected position 24812212 , only wrote 24811920
2019-03-21 12:48:41.540 1539-1652/? I/Netd: Destroyed 7 sockets on 192.168.232.2 in 1.7 ms
2019-03-21 12:48:41.540 2194-20570/com.google.android.gms.persistent V/NativeCrypto: Read error: ssl=0xaf22b998: I/O error during system call, Software caused connection abort
2019-03-21 12:48:41.543 1680-1756/system_process E/WifiVendorHal: stopRssiMonitoring(l.2115) failed {.code = ERROR_NOT_AVAILABLE, .description = }
2019-03-21 12:48:41.543 1680-1756/system_process W/WifiConfigManager: Looking up network with invalid networkId -1
2019-03-21 12:48:41.544 2194-20570/com.google.android.gms.persistent V/NativeCrypto: SSL shutdown failed: ssl=0xaf22b998: I/O error during system call, Broken pipe
2019-03-21 12:48:41.544 2194-20570/com.google.android.gms.persistent E/WakeLock: GCM_HB_ALARM release without a matched acquire!
2019-03-21 12:48:41.546 20467-20498/fr.free.nrw.commons.beta I/DefaultRequestDirector: I/O exception (javax.net.ssl.SSLException) caught when processing request: Write error: ssl=0xa1722780: I/O error during system call, Software caused connection abort
2019-03-21 12:48:41.546 1680-1760/system_process D/ConnectivityService: NetworkAgentInfo [WIFI () - 105] EVENT_NETWORK_INFO_CHANGED, going from CONNECTED to DISCONNECTED
2019-03-21 12:48:41.546 1680-1760/system_process D/ConnectivityService: NetworkAgentInfo [WIFI () - 105] got DISCONNECTED, was satisfying 12
2019-03-21 12:48:41.547 20467-20498/fr.free.nrw.commons.beta I/DefaultRequestDirector: Retrying request
2019-03-21 12:48:41.547 2194-20570/com.google.android.gms.persistent E/GCM: Connection closed with errorCode 20
2019-03-21 12:48:41.569 2194-2194/com.google.android.gms.persistent I/GeofencerStateMachine: sendWifiConnectivityChanged: connectivity=false
2019-03-21 12:48:41.570 1539-1660/? D/CommandListener: Clearing all IP addresses on wlan0
2019-03-21 12:48:41.573 1680-1756/system_process D/WifiCountryCode: Succeeded to set country code to: US
2019-03-21 12:48:41.575 1667-1667/? I/hostapd: wlan1: AP-STA-DISCONNECTED 02:00:00:44:55:66
2019-03-21 12:48:41.578 20467-20496/fr.free.nrw.commons.beta W/System.err: io.reactivex.exceptions.CompositeException: 2 exceptions occurred. 
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.observers.ConsumerSingleObserver.onError(ConsumerSingleObserver.java:49)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onError(SingleFlatMap.java:90)
2019-03-21 12:48:41.583 2194-20570/com.google.android.gms.persistent E/WakeLock: GCM_CONN_ALARM release without a matched acquire!
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:79)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
2019-03-21 12:48:41.583 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.lang.Thread.run(Thread.java:764)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:   ComposedException 1 :
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err: 	java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.beta.wmflabs.org": No address associated with hostname
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:141)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:90)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.net.InetAddress.getAllByName(InetAddress.java:787)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:142)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:169)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:124)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:466)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:560)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:492)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:470)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at in.yuvi.http.fluent.Http$HttpRequestBuilder.asResponse(Http.java:470)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.CustomApiResult.fromRequestBuilder(CustomApiResult.java:42)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.CustomMwApi.uploadToStash(CustomMwApi.java:149)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApi.lambda$uploadFile$3(ApacheHttpClientMediaWikiApi.java:864)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.-$$Lambda$ApacheHttpClientMediaWikiApi$xiJwLUpZnd3fpDlWkWsXe3p1svQ.call(Unknown Source:10)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.Single.subscribe(Single.java:3438)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2019-03-21 12:48:41.584 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.lang.Thread.run(Thread.java:764)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err: 	Caused by: android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at libcore.io.Linux.android_getaddrinfo(Native Method)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:58)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:122)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:90)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.net.InetAddress.getAllByName(InetAddress.java:787)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:142)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:169)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:124)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:466)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:560)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:492)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:470)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at in.yuvi.http.fluent.Http$HttpRequestBuilder.asResponse(Http.java:470)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.CustomApiResult.fromRequestBuilder(CustomApiResult.java:42)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.CustomMwApi.uploadToStash(CustomMwApi.java:149)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApi.lambda$uploadFile$3(ApacheHttpClientMediaWikiApi.java:864)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.mwapi.-$$Lambda$ApacheHttpClientMediaWikiApi$xiJwLUpZnd3fpDlWkWsXe3p1svQ.call(Unknown Source:10)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
2019-03-21 12:48:41.585 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.Single.subscribe(Single.java:3438)
2019-03-21 12:48:41.586 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
2019-03-21 12:48:41.586 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
2019-03-21 12:48:41.586 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:     at java.lang.Thread.run(Thread.java:764)
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:   ComposedException 2 :
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err: 	java.lang.RuntimeException: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.beta.wmflabs.org": No address associated with hostname
2019-03-21 12:48:41.594 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.upload.UploadService.lambda$uploadContribution$2(UploadService.java:295)
2019-03-21 12:48:41.595 20467-20496/fr.free.nrw.commons.beta W/System.err:     at fr.free.nrw.commons.upload.-$$Lambda$UploadService$kRDkGIKcRgpB3TjwhqZA9xAAnWE.accept(Unknown Source:2)
2019-03-21 12:48:41.597 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.observers.ConsumerSingleObserver.onError(ConsumerSingleObserver.java:46)
2019-03-21 12:48:41.597 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onError(SingleFlatMap.java:90)
2019-03-21 12:48:41.597 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:79)
2019-03-21 12:48:41.597 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
2019-03-21 12:48:41.597 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
2019-03-21 12:48:41.597 20467-20496/fr.free.nrw.commons.beta W/System.err:     at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
2019-03-21 12:48:41.597 20467-20496/fr.f
```

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
@@ -944,7 +944,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             long dataLength,
             Uri fileUri,
             Uri contentProviderUri,
-            ProgressListener progressListener) throws IOException {
+            ProgressListener progressListener) {
         return Single.fromCallable(() -> {
             CustomApiResult result = api.uploadToStash(filename, file, dataLength, getEditToken(), progressListener::onProgress);
 

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
@@ -9,7 +9,6 @@ import java.io.InputStream;
 import java.util.List;
 
 import fr.free.nrw.commons.Media;
-import fr.free.nrw.commons.campaigns.CampaignResponseDTO;
 import fr.free.nrw.commons.notification.Notification;
 import io.reactivex.Observable;
 import io.reactivex.Single;
@@ -53,7 +52,7 @@ public interface MediaWikiApi {
     @NonNull
     Single<UploadStash> uploadFile(String filename, InputStream file,
                                    long dataLength, Uri fileUri, Uri contentProviderUri,
-                                   final ProgressListener progressListener) throws IOException;
+                                   final ProgressListener progressListener);
 
     @NonNull
     Single<UploadResult> uploadFileFinalize(String filename, String filekey,

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
@@ -36,7 +36,6 @@ import fr.free.nrw.commons.contributions.MainActivity;
 import fr.free.nrw.commons.mwapi.MediaWikiApi;
 import fr.free.nrw.commons.wikidata.WikidataEditService;
 import io.reactivex.Single;
-import io.reactivex.functions.Action;
 import io.reactivex.schedulers.Schedulers;
 import timber.log.Timber;
 
@@ -196,6 +195,7 @@ public class UploadService extends HandlerService<Contribution> {
                 .setContentIntent(PendingIntent.getActivity(this, 0, new Intent(this, MainActivity.class), 0));
     }
 
+    @SuppressLint("CheckResult")
     private void uploadContribution(Contribution contribution) {
         InputStream fileInputStream;
         Uri localUri = contribution.getLocalUri();
@@ -223,7 +223,13 @@ public class UploadService extends HandlerService<Contribution> {
 
         String filename = contribution.getFilename();
 
-        try {
+        NotificationUpdateProgressListener notificationUpdater = new NotificationUpdateProgressListener(notificationTag,
+                getString(R.string.upload_progress_notification_title_in_progress, contribution.getDisplayTitle()),
+                getString(R.string.upload_progress_notification_title_finishing, contribution.getDisplayTitle()),
+                contribution
+        );
+
+        Single.fromCallable(() -> {
             if (!mwApi.validateLogin()) {
                 // Need to revalidate!
                 if (sessionManager.revalidateAuthToken()) {
@@ -231,31 +237,27 @@ public class UploadService extends HandlerService<Contribution> {
                 } else {
                     Timber.d("Unable to revalidate :(");
                     stopForeground(true);
-                    Toast failureToast = Toast.makeText(this, R.string.authentication_failed, Toast.LENGTH_LONG);
-                    failureToast.show();
-                    sessionManager.forceLogin(this);
-                    return;
+                    sessionManager.forceLogin(UploadService.this);
+                    throw new RuntimeException(getString(R.string.authentication_failed));
                 }
             }
-        } catch (IOException e) {
-            onUploadError(contribution, e);
-            onUploadFinally(filename);
-            return;
-        }
-
-        NotificationUpdateProgressListener notificationUpdater = new NotificationUpdateProgressListener(notificationTag,
-                getString(R.string.upload_progress_notification_title_in_progress, contribution.getDisplayTitle()),
-                getString(R.string.upload_progress_notification_title_finishing, contribution.getDisplayTitle()),
-                contribution
-        );
-
-        String stashFilename = "Temp_" + contribution.hashCode() + filename;
-        mwApi.uploadFile(
+            return "Temp_" + contribution.hashCode() + filename;
+        }).flatMap(stashFilename -> mwApi.uploadFile(
                 stashFilename, fileInputStream, contribution.getDataLength(),
-                localUri, contribution.getContentProviderUri(), notificationUpdater)
+                localUri, contribution.getContentProviderUri(), notificationUpdater))
                 .subscribeOn(Schedulers.io())
                 .observeOn(Schedulers.io())
-                .doFinally(() -> onUploadFinally(filename))
+                .doFinally(() -> {
+                    if (filename != null) {
+                        unfinishedUploads.remove(filename);
+                    }
+                    toUpload--;
+                    if (toUpload == 0) {
+                        // Sync modifications right after all uploads are processed
+                        ContentResolver.requestSync(sessionManager.getCurrentAccount(), BuildConfig.MODIFICATION_AUTHORITY, new Bundle());
+                        stopForeground(true);
+                    }
+                })
                 .flatMap(uploadStash -> {
                     notificationManager.cancel(NOTIFICATION_UPLOAD_IN_PROGRESS);
 
@@ -299,25 +301,11 @@ public class UploadService extends HandlerService<Contribution> {
                         contribution.setDateUploaded(uploadResult.getDateUploaded());
                         contributionDao.save(contribution);
                     }
-                }, throwable -> onUploadError(contribution, throwable));
-    }
-
-    private void onUploadError(Contribution contribution, Throwable t) {
-        Timber.w(t,"Exception during upload");
-        notificationManager.cancel(NOTIFICATION_UPLOAD_IN_PROGRESS);
-        showFailedNotification(contribution);
-    }
-
-    private void onUploadFinally(String filename) {
-        if (filename != null) {
-            unfinishedUploads.remove(filename);
-        }
-        toUpload--;
-        if (toUpload == 0) {
-            // Sync modifications right after all uploads are processed
-            ContentResolver.requestSync(sessionManager.getCurrentAccount(), BuildConfig.MODIFICATION_AUTHORITY, new Bundle());
-            stopForeground(true);
-        }
+                }, throwable -> {
+                    Timber.w(throwable, "Exception during upload");
+                    notificationManager.cancel(NOTIFICATION_UPLOAD_IN_PROGRESS);
+                    showFailedNotification(contribution);
+                });
     }
 
     @SuppressLint("StringFormatInvalid")

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
@@ -36,6 +36,7 @@ import fr.free.nrw.commons.contributions.MainActivity;
 import fr.free.nrw.commons.mwapi.MediaWikiApi;
 import fr.free.nrw.commons.wikidata.WikidataEditService;
 import io.reactivex.Single;
+import io.reactivex.functions.Action;
 import io.reactivex.schedulers.Schedulers;
 import timber.log.Timber;
 
@@ -223,7 +224,6 @@ public class UploadService extends HandlerService<Contribution> {
         String filename = contribution.getFilename();
 
         try {
-
             if (!mwApi.validateLogin()) {
                 // Need to revalidate!
                 if (sessionManager.revalidateAuthToken()) {
@@ -237,77 +237,86 @@ public class UploadService extends HandlerService<Contribution> {
                     return;
                 }
             }
-            NotificationUpdateProgressListener notificationUpdater = new NotificationUpdateProgressListener(notificationTag,
-                    getString(R.string.upload_progress_notification_title_in_progress, contribution.getDisplayTitle()),
-                    getString(R.string.upload_progress_notification_title_finishing, contribution.getDisplayTitle()),
-                    contribution
-            );
-            String stashFilename = "Temp_" + contribution.hashCode() + filename;
-            mwApi.uploadFile(
-                    stashFilename, fileInputStream, contribution.getDataLength(),
-                    localUri, contribution.getContentProviderUri(), notificationUpdater)
-                    .subscribeOn(Schedulers.io())
-                    .observeOn(Schedulers.io())
-                    .flatMap(uploadStash -> {
-                        notificationManager.cancel(NOTIFICATION_UPLOAD_IN_PROGRESS);
-
-                        Timber.d("Stash upload response 1 is %s", uploadStash.toString());
-
-                        String resultStatus = uploadStash.getResultStatus();
-                        if (!resultStatus.equals("Success")) {
-                            Timber.d("Contribution upload failed. Wikidata entity won't be edited");
-                            showFailedNotification(contribution);
-                            return Single.never();
-                        } else {
-                            synchronized (unfinishedUploads) {
-                                Timber.d("making sure of uniqueness of name: %s", filename);
-                                String uniqueFilename = findUniqueFilename(filename);
-                                unfinishedUploads.add(uniqueFilename);
-                                return mwApi.uploadFileFinalize(
-                                        uniqueFilename,
-                                        uploadStash.getFilekey(),
-                                        contribution.getPageContents(getApplicationContext()),
-                                        contribution.getEditSummary());
-                            }
-                        }
-                    })
-                    .subscribe(uploadResult -> {
-                        Timber.d("Stash upload response 2 is %s", uploadResult.toString());
-
-                        notificationManager.cancel(notificationTag, NOTIFICATION_UPLOAD_IN_PROGRESS);
-
-                        String resultStatus = uploadResult.getResultStatus();
-                        if (!resultStatus.equals("Success")) {
-                            Timber.d("Contribution upload failed. Wikidata entity won't be edited");
-                            showFailedNotification(contribution);
-                        } else {
-                            String canonicalFilename = uploadResult.getCanonicalFilename();
-                            Timber.d("Contribution upload success. Initiating Wikidata edit for entity id %s",
-                                    contribution.getWikiDataEntityId());
-                            wikidataEditService.createClaimWithLogging(contribution.getWikiDataEntityId(), canonicalFilename);
-                            contribution.setFilename(canonicalFilename);
-                            contribution.setImageUrl(uploadResult.getImageUrl());
-                            contribution.setState(Contribution.STATE_COMPLETED);
-                            contribution.setDateUploaded(uploadResult.getDateUploaded());
-                            contributionDao.save(contribution);
-                        }
-                    }, throwable -> {
-                        throw new RuntimeException(throwable);
-                    });
         } catch (IOException e) {
-            Timber.w(e,"IOException during upload");
-            notificationManager.cancel(NOTIFICATION_UPLOAD_IN_PROGRESS);
-            showFailedNotification(contribution);
-        } finally {
-            if (filename != null) {
-                unfinishedUploads.remove(filename);
-            }
-            toUpload--;
-            if (toUpload == 0) {
-                // Sync modifications right after all uplaods are processed
-                ContentResolver.requestSync(sessionManager.getCurrentAccount(), BuildConfig.MODIFICATION_AUTHORITY, new Bundle());
-                stopForeground(true);
-            }
+            onUploadError(contribution, e);
+            onUploadFinally(filename);
+            return;
+        }
+
+        NotificationUpdateProgressListener notificationUpdater = new NotificationUpdateProgressListener(notificationTag,
+                getString(R.string.upload_progress_notification_title_in_progress, contribution.getDisplayTitle()),
+                getString(R.string.upload_progress_notification_title_finishing, contribution.getDisplayTitle()),
+                contribution
+        );
+
+        String stashFilename = "Temp_" + contribution.hashCode() + filename;
+        mwApi.uploadFile(
+                stashFilename, fileInputStream, contribution.getDataLength(),
+                localUri, contribution.getContentProviderUri(), notificationUpdater)
+                .subscribeOn(Schedulers.io())
+                .observeOn(Schedulers.io())
+                .doFinally(() -> onUploadFinally(filename))
+                .flatMap(uploadStash -> {
+                    notificationManager.cancel(NOTIFICATION_UPLOAD_IN_PROGRESS);
+
+                    Timber.d("Stash upload response 1 is %s", uploadStash.toString());
+
+                    String resultStatus = uploadStash.getResultStatus();
+                    if (!resultStatus.equals("Success")) {
+                        Timber.d("Contribution upload failed. Wikidata entity won't be edited");
+                        showFailedNotification(contribution);
+                        return Single.never();
+                    } else {
+                        synchronized (unfinishedUploads) {
+                            Timber.d("making sure of uniqueness of name: %s", filename);
+                            String uniqueFilename = findUniqueFilename(filename);
+                            unfinishedUploads.add(uniqueFilename);
+                            return mwApi.uploadFileFinalize(
+                                    uniqueFilename,
+                                    uploadStash.getFilekey(),
+                                    contribution.getPageContents(getApplicationContext()),
+                                    contribution.getEditSummary());
+                        }
+                    }
+                })
+                .subscribe(uploadResult -> {
+                    Timber.d("Stash upload response 2 is %s", uploadResult.toString());
+
+                    notificationManager.cancel(notificationTag, NOTIFICATION_UPLOAD_IN_PROGRESS);
+
+                    String resultStatus = uploadResult.getResultStatus();
+                    if (!resultStatus.equals("Success")) {
+                        Timber.d("Contribution upload failed. Wikidata entity won't be edited");
+                        showFailedNotification(contribution);
+                    } else {
+                        String canonicalFilename = uploadResult.getCanonicalFilename();
+                        Timber.d("Contribution upload success. Initiating Wikidata edit for entity id %s",
+                                contribution.getWikiDataEntityId());
+                        wikidataEditService.createClaimWithLogging(contribution.getWikiDataEntityId(), canonicalFilename);
+                        contribution.setFilename(canonicalFilename);
+                        contribution.setImageUrl(uploadResult.getImageUrl());
+                        contribution.setState(Contribution.STATE_COMPLETED);
+                        contribution.setDateUploaded(uploadResult.getDateUploaded());
+                        contributionDao.save(contribution);
+                    }
+                }, throwable -> onUploadError(contribution, throwable));
+    }
+
+    private void onUploadError(Contribution contribution, Throwable t) {
+        Timber.w(t,"Exception during upload");
+        notificationManager.cancel(NOTIFICATION_UPLOAD_IN_PROGRESS);
+        showFailedNotification(contribution);
+    }
+
+    private void onUploadFinally(String filename) {
+        if (filename != null) {
+            unfinishedUploads.remove(filename);
+        }
+        toUpload--;
+        if (toUpload == 0) {
+            // Sync modifications right after all uploads are processed
+            ContentResolver.requestSync(sessionManager.getCurrentAccount(), BuildConfig.MODIFICATION_AUTHORITY, new Bundle());
+            stopForeground(true);
         }
     }
 

